In [1]:
import os 

hypos = []
for root, dirs, files in os.walk("hypos"):
    for file in files:
        if file.startswith("hypo") and file.endswith(".txt"):
            with open(os.path.join(root, file)) as f:
                hypos.append([line.strip() for line in f])
                
ground_truths = []
for root, dirs, files in os.walk("ground_truths"):
    for file in files:
        if file.startswith("gt") and file.endswith(".txt"):
            with open(os.path.join(root, file)) as f:
                ground_truths.append([line.strip() for line in f])

In [2]:
len(hypos), len(ground_truths)

(5, 3)

In [3]:
import unicodedata
import re

def read_vi_text(text):
    text = unicodedata.normalize("NFC", text)
    text = re.sub(r'\s+', ' ', text)
    return text

read_vi = []
for i, gt in enumerate(ground_truths):
    ground_truths[i] = [read_vi_text(line) for line in gt]
    read_vi.append(ground_truths[i])

In [4]:
read_vi

[['1. Giới thiệu chung',
  'Ngành Kỹ thuật Điện tử – Viễn thông là lĩnh vực nghiên cứu và ứng dụng các công nghệ tiên tiến nhất để phát triển các thiết bị điện tử và hệ thống viễn thông. Trong ngành này, có hai lĩnh vực chính là Điện tử và Viễn thông. Lĩnh vực Điện tử tập trung vào nghiên cứu và sản xuất vi mạch điện tử, là bộ não điều khiển của các thiết bị thông minh. Lĩnh vực Viễn thông tập trung vào nghiên cứu và sử dụng các thiết bị tạo nên hệ thống mạng viễn thông, bao gồm các thành phần như thiết bị truyền dẫn, thiết bị chuyển mạch, thiết bị đầu cuối và môi trường truyền. Ngành Kỹ thuật Điện tử - Viễn thông đóng vai trò quan trọng trong viêc giúp xây dựng và duy trì các hệ thống mạng thông tin liên lạc toàn cầu, đảm bảo việc trao đổi thông tin diễn ra thuận lợi mọi lúc, mọi nơi.',
  '2. Chương trình đào tạo',
  '1. Mục tiêu chương trình',
  'Chương trình nhằm trang bị sinh viên với kiến thức và kỹ năng toàn diện về lĩnh vực Điện tử - Viễn thông. Sinh viên sẽ được phát triển về m

In [5]:
import textwrap

for ref in read_vi[0]:
    print(textwrap.fill(ref, width=120))
    print()

1. Giới thiệu chung

Ngành Kỹ thuật Điện tử – Viễn thông là lĩnh vực nghiên cứu và ứng dụng các công nghệ tiên tiến nhất để phát triển các
thiết bị điện tử và hệ thống viễn thông. Trong ngành này, có hai lĩnh vực chính là Điện tử và Viễn thông. Lĩnh vực Điện
tử tập trung vào nghiên cứu và sản xuất vi mạch điện tử, là bộ não điều khiển của các thiết bị thông minh. Lĩnh vực Viễn
thông tập trung vào nghiên cứu và sử dụng các thiết bị tạo nên hệ thống mạng viễn thông, bao gồm các thành phần như
thiết bị truyền dẫn, thiết bị chuyển mạch, thiết bị đầu cuối và môi trường truyền. Ngành Kỹ thuật Điện tử - Viễn thông
đóng vai trò quan trọng trong viêc giúp xây dựng và duy trì các hệ thống mạng thông tin liên lạc toàn cầu, đảm bảo việc
trao đổi thông tin diễn ra thuận lợi mọi lúc, mọi nơi.

2. Chương trình đào tạo

1. Mục tiêu chương trình

Chương trình nhằm trang bị sinh viên với kiến thức và kỹ năng toàn diện về lĩnh vực Điện tử - Viễn thông. Sinh viên sẽ
được phát triển về mặt chuyên môn, kỹ n

In [6]:
from unidecode import unidecode
import re

def normalize_text(text):
    """
    Normalize Vietnamese text by converting it to its ASCII representation and removing accents.
    
    Args:
    text (str): Input Vietnamese text to be normalized.
    
    Returns:
    str: Normalized text with accents removed.
    """
    normalized_text = unidecode(text)
    normalized_text = re.sub(r'\s+', ' ', normalized_text).strip().lower()
    return normalized_text


In [7]:
normalized_hypos = [[' '.join(normalize_text(text) for text in hypo)] for hypo in hypos]

In [8]:
len(normalized_hypos)

5

In [9]:
normalized_ground_truths = [[' '.join(normalize_text(text) for text in ground_truth)] for ground_truth in ground_truths]

In [10]:
len(normalized_ground_truths)

3

In [11]:
import textwrap

def wraptext(documents, width=120):
    for idx, doc in enumerate(documents, start=1):
        wrapped_lines = textwrap.wrap(doc, width=width)
        for line in wrapped_lines:
            print(line)
        print("-" * width)

In [12]:
for i in range(len(normalized_ground_truths)):
    print(f"Document {i + 1}:")
    wraptext(normalized_ground_truths[i])

Document 1:
1. gioi thieu chung nganh ky thuat dien tu - vien thong la linh vuc nghien cuu va ung dung cac cong nghe tien tien nhat
de phat trien cac thiet bi dien tu va he thong vien thong. trong nganh nay, co hai linh vuc chinh la dien tu va vien
thong. linh vuc dien tu tap trung vao nghien cuu va san xuat vi mach dien tu, la bo nao dieu khien cua cac thiet bi
thong minh. linh vuc vien thong tap trung vao nghien cuu va su dung cac thiet bi tao nen he thong mang vien thong, bao
gom cac thanh phan nhu thiet bi truyen dan, thiet bi chuyen mach, thiet bi dau cuoi va moi truong truyen. nganh ky thuat
dien tu - vien thong dong vai tro quan trong trong viec giup xay dung va duy tri cac he thong mang thong tin lien lac
toan cau, dam bao viec trao doi thong tin dien ra thuan loi moi luc, moi noi. 2. chuong trinh dao tao 1. muc tieu chuong
trinh chuong trinh nham trang bi sinh vien voi kien thuc va ky nang toan dien ve linh vuc dien tu - vien thong. sinh
vien se duoc phat trien ve mat chuyen m

In [13]:
for i in range(len(normalized_hypos)):
    print(f"Document {i + 1}:")
    wraptext(normalized_hypos[i])

Document 1:
# tong quan ve nganh dien tu - vien thong cua dai hoc bach khoa ha noi  nganh dien tu - vien thong cua dai hoc bach khoa
ha noi (hust) la mot trong nhung nganh hoc chu chot, dap ung nhu cau nhan luc chat luong cao trong thoi dai cong nghe so
va cach mang cong nghiep 4.0. duoc biet den voi chuong trinh dao tao tien tien, nganh nay khong chi thu hut su quan tam
cua sinh vien trong nuoc ma con cua ca sinh vien quoc te.  ## chuong trinh dao tao  chuong trinh dao tao nganh dien tu -
vien thong tai hust duoc thiet ke de cung cap cho sinh vien mot nen tang vung chac ve toan hoc va khoa hoc co ban, cung
voi khoi kien thuc co so nganh va kien thuc chuyen sau theo cac dinh huong ky thuat khac nhau. sinh vien co co hoi lua
chon chuyen nganh phu hop voi so thich va nhu cau nghe nghiep, bao gom dien tu va ky thuat may tinh, ky thuat thong tin
- truyen thong, ky thuat y sinh, ky thuat dien tu hang khong vu tru, va ky thuat da phuong tien (hocmai.vn, 2024).
chuong trinh duoc giang day ban

### Using BERTScore for Evaluation

1. BERTScore is used to measure textual similarity between candidate texts and reference texts. It considers not only exact word matches but also the overall meaning, fluency, and order of the ouput.

2. BERTScore: Precision, Recall, F1
    * Precision measures how well the candidate texts avoid introducing irrelevant content.
    * Recall measures how well the candidate texts avoid omitting relevant content.
    * F1 = 2 x (P x R)/(P + R) 

In [14]:
# Hide the loading messages
import logging
import transformers
transformers.tokenization_utils.logger.setLevel(logging.ERROR)
transformers.configuration_utils.logger.setLevel(logging.ERROR)
transformers.modeling_utils.logger.setLevel(logging.ERROR)

/media/sanslab/Data/stevehoang/rag_research/vigpt-researcher/venv/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [15]:
from bert_score import score

In [16]:
def bert_score(hypos, refs, lang="vi"):
    bert_scores = []
    for hypo in hypos:
        for ref in refs:
            scores = score(hypo, ref, lang=lang, verbose=False)
            bert_scores.append(scores)
    return bert_scores

In [17]:
bert_scores = bert_score(normalized_hypos, normalized_ground_truths)

In [18]:
import numpy as np

bert_scores = np.array(bert_scores)

print(f"BERTScore:")
print(f"Precision: {bert_scores[:, 0].mean():.2f}")
print(f"Recall: {bert_scores[:, 1].mean():.2f}")
print(f"F1: {bert_scores[:, 2].mean():.2f}")

BERTScore:
Precision: 0.81
Recall: 0.83
F1: 0.82


### Using ROUGE score for Evaluation

In [19]:
from rouge_score import rouge_scorer

def rouge_score(hypos, refs):
    scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)
    scores = [scorer.score(ref, hypo) for ref, hypo in zip(refs, hypos)]
    return scores

rouge_scores = []
for i, (hypo, ref) in enumerate(zip(normalized_hypos, normalized_ground_truths), start=1):
    rouge_scores.append(rouge_score(hypo, ref))

In [20]:
rouge_1_precisions = []
rouge_1_recalls = []
rouge_1_fmeasures = []
rouge_2_precisions = []
rouge_2_recalls = []
rouge_2_fmeasures = []
rouge_L_precisions = []
rouge_L_recalls = []
rouge_L_fmeasures = []

for rouge_score in rouge_scores:
    for scores in rouge_score:
        rouge_1 = scores['rouge1']
        rouge_2 = scores['rouge2']
        rouge_L = scores['rougeL']
        
        rouge_1_precisions.append(rouge_1.precision)
        rouge_1_recalls.append(rouge_1.recall)
        rouge_1_fmeasures.append(rouge_1.fmeasure)
        rouge_2_precisions.append(rouge_2.precision)
        rouge_2_recalls.append(rouge_2.recall)
        rouge_2_fmeasures.append(rouge_2.fmeasure)
        rouge_L_precisions.append(rouge_L.precision)
        rouge_L_recalls.append(rouge_L.recall)
        rouge_L_fmeasures.append(rouge_L.fmeasure)

In [21]:
rouge_1_precisions = np.array(rouge_1_precisions)
print(f"ROUGE-1 Precision: {rouge_1_precisions.mean():.2f}")
rouge_1_recalls = np.array(rouge_1_recalls)
print(f"ROUGE-1 Recall: {rouge_1_recalls.mean():.2f}")
rouge_1_fmeasures = np.array(rouge_1_fmeasures)
print(f"ROUGE-1 F1: {rouge_1_fmeasures.mean():.2f}")
rouge_2_precisions = np.array(rouge_2_precisions)
print(f"ROUGE-2 Precision: {rouge_2_precisions.mean():.2f}")
rouge_2_recalls = np.array(rouge_2_recalls)
print(f"ROUGE-2 Recall: {rouge_2_recalls.mean():.2f}")
rouge_2_fmeasures = np.array(rouge_2_fmeasures)
print(f"ROUGE-2 F1: {rouge_2_fmeasures.mean():.2f}")
rouge_L_precisions = np.array(rouge_L_precisions)
print(f"ROUGE-L Precision: {rouge_L_precisions.mean():.2f}")
rouge_L_recalls = np.array(rouge_L_recalls)
print(f"ROUGE-L Recall: {rouge_L_recalls.mean():.2f}")
rouge_L_fmeasures = np.array(rouge_L_fmeasures)
print(f"ROUGE-L F1: {rouge_L_fmeasures.mean():.2f}")

ROUGE-1 Precision: 0.82
ROUGE-1 Recall: 0.50
ROUGE-1 F1: 0.61
ROUGE-2 Precision: 0.46
ROUGE-2 Recall: 0.28
ROUGE-2 F1: 0.35
ROUGE-L Precision: 0.40
ROUGE-L Recall: 0.25
ROUGE-L F1: 0.30


### Using Perplexity for Evaluation

In [22]:
# Define the function to calculate Perplexity score
from transformers import BertTokenizer, BertForMaskedLM
import torch


def perplexity_score(hypos, tokenizer, model):
    perplexity_scores = []
    for hypo in hypos:
        input_ids = tokenizer.encode(hypo[0], return_tensors="pt")
        # Truncate the sequence if it's longer than the model's maximum input length
        if input_ids.size(1) > 512:
            input_ids = input_ids[:, :512]
        with torch.no_grad():
            loss = model(input_ids, labels=input_ids)[0]
        perplexity = torch.exp(loss).item()
        perplexity_scores.append(perplexity)
    return perplexity_scores

# Load the pre-trained BERT model
tokenizer = BertTokenizer.from_pretrained("bert-base-multilingual-cased")
model = BertForMaskedLM.from_pretrained("bert-base-multilingual-cased")

# Calculate the Perplexity score
perplexity = perplexity_score(normalized_hypos, tokenizer, model)

Token indices sequence length is longer than the specified maximum sequence length for this model (1067 > 512). Running this sequence through the model will result in indexing errors


In [23]:
perplexity = np.array(perplexity)

In [24]:
print(f"Perplexity: {perplexity.mean():.2f}")

Perplexity: 1.06


### RESULTS

| METRIC   | Precision | Recall | F1 |
|---------|----------|-------|-------|
| BERTScore|0.81 |0.83 |0.82 |
| ROUGE-1| 0.82|0.50 |0.61 |
| ROUGE-2| 0.46|0.28 | 0.35|
| ROUGE-L| 0.40|0.25 |0.30 |

Perlexity = 1.06

### Total Time Executions

1. Câu hỏi 1: Tổng quan về ngành Điện tử - Viễn thông của Đại học Bách Khoa Hà Nội
2. Câu hỏi 2: Trình bày cơ chế hoạt động của GPT-4
3. Câu hỏi 3: Cuộc đời của Steve Jobs như thế nào?
4. Câu hỏi 4: Bạn có thể cho tôi một đoạn thơ ngắn để tỏ tình crush được không?
5. Câu hỏi 5: Sự phát triển của chíp bán dẫn ở Việt Nam như thế nào?

#### Đối với nhiệm vụ viết báo cáo nghiên cứu

|Time (s) | T1 | T2 |
|----|----|----|
|    | |  |

#### Đối với nhiệm vụ phân tích nguồn tham khảo

|Time (s) | T1 | T2 |
|----|----|----|
|    |  | 123.123 |

#### Đôi với nhiệm vụ viết khung báo cáo 

|Time (s) | T1 | T2 |
|----|----|----|
|    |  | 123.123 |

#### Đối với nhiệm vụ viết câu trả lời cho câu hỏi

|Time (s) | T1 | T2 |
|----|----|----|
|    | 100.123 | 123.123 |

### Review
1. Summarization is a complex task, even advanced models can struggle to accurately capture all the important information from a source text.

2. BERTScore or ROUGE that these metrics focus on things like including keywords from the source text, so that do not perfectly reflect human judgement of a good summary.

3. The response of chatbot from many different sources.


### HUMAN EVALUATION IS THE BEST WAY TO EVALUATE THE QUALITY OF A SUMMARY